ru/lessons/basics/functions.md

```diff
@@ -0,0 +1,216 @@
+
+iex> Greeter.hello("Sean")
+"Hello, Sean"
+```
+
+Если функция однострочная, то ее описание можно сократить с использованием `do:`:
+
+```elixir
+defmodule Greeter do
+  def hello(name), do: "Hello, " <> name
+end
+```
+
+Уже разобравшись в сопоставлении с образцом, давайте рассмотрим пример рекурсии с использованием именованных функций:
+
+```elixir
+defmodule Length do
+  def of([]), do: 0
+  def of([_|t]), do: 1 + of(t)
+end
+
+iex> Length.of []
+0
+iex> Length.of [1, 2, 3]
+3
+```
+
+### Закрытые функции
+
+Когда мы не хотим давать доступ к функции из других модулей, мы определяем закрытые (private) функции. Они могут быть вызваны только из этого же модуля. Такие функции определяются с помощью `defp`:
+
+```elixir
+defmodule Greeter do
+  def hello(name), do: phrase <> name
+  defp phrase, do: "Hello, "
+end
+
+iex> Greeter.hello("Sean")
+"Hello, Sean"
+
+iex> Greeter.phrase
+** (UndefinedFunctionError) undefined function: Greeter.phrase/0
+    Greeter.phrase()
+```
+
+### Ограничители
+
+Мы уже затрагивали ограничители в главе [Управляющие конструкции](../control-structures.md), теперь же рассмотрим их применение в именованных функциях. Ограничители проверяются только после того как только Elixir сопоставил функцию.
+
+В следующем примере у нас есть две функции с одинаковыми сигнатурами. Мы используем ограничители для определения какую из них использовать на основе типа аргумента:
+
+```elixir
+defmodule Greeter do
+  def hello(names) when is_list(names) do
+    names
+    |> Enum.join(", ")
+    |> hello
+  end
+
+  def hello(name) when is_binary(name) do
+    phrase <> name
+  end
+
+  defp phrase, do: "Hello, "
+end
+
+iex> Greeter.hello ["Sean", "Steve"]
+"Hello, Sean, Steve"
+```
+
+### Аргументы по умолчанию
+
+Когда мы хотим иметь некое значение по умолчанию у аргумента - используется синтаксис `argument \\ value`:
+
+```elixir
+defmodule Greeter do
+  def hello(name, country \\ "en") do
+    phrase(country) <> name
+  end
+
+  defp phrase("en"), do: "Hello, "
+  defp phrase("es"), do: "Hola, "
+end
+
+iex> Greeter.hello("Sean", "en")
+"Hello, Sean"
+
+iex> Greeter.hello("Sean")
+"Hello, Sean"
+
+iex> Greeter.hello("Sean", "es")
+"Hola, Sean"
+```
+
+Когда мы используем одновременно ограничители и аргументы по умолчанию, то все перестает работать. Давайте посмотрим как это выглядит:
+
+```elixir
+defmodule Greeter do
+  def hello(names, country \\ "en") when is_list(names) do
+    names
+    |> Enum.join(", ")
+    |> hello(country)
+  end
+
+  def hello(name, country \\ "en") when is_binary(name) do
+    phrase(country) <> name
+  end
+
+  defp phrase("en"), do: "Hello, "
+  defp phrase("es"), do: "Hola, "
+end
+
+** (CompileError) def hello/2 has default values and multiple clauses, define a function head with the defaults
+```
+
+Elixir не поддерживает аргументы по умолчанию при наличии нескольких подходящих функций, ведь это может запутать. Для решения этой проблемы мы добавляем определение функции с аргументами по умолчанию:
```

Comment:
Well... I don't like it too. This part can be thrown out, but then we're moving more and more to literary translation.